File: src/main/java/io/github/svegon/capi/event/render/NauseaRenderCallback.java
package io.github.svegon.capi.event.render;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.client.gui.DrawContext;
import net.minecraft.client.render.GameRenderer;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

@FunctionalInterface
public interface NauseaRenderCallback {
    Event<NauseaRenderCallback> EVENT = EventFactory.createArrayBacked(NauseaRenderCallback.class,
            (gameRenderer, context, distortionStrength, callback) -> {},
            listeners -> (gameRenderer, context, distortionStrength, callback) -> {
        for (NauseaRenderCallback listener : listeners) {
            listener.onNauseaRender(gameRenderer, context, distortionStrength, callback);

            if (callback.isCancelled()) {
                return;
            }
        }
    });

    void onNauseaRender(GameRenderer gameRenderer, DrawContext context, float distortionStrength,
                        CallbackInfo callback);
}

File: src/client/kotlin/io/github/svegon/mclientapi/client/event/render/NauseaRenderCallback.kt
package io.github.svegon.mclientapi.client.event.render

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.client.gui.DrawContext
import net.minecraft.client.render.GameRenderer
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo

fun interface NauseaRenderCallback {
    fun onNauseaRender(
        gameRenderer: GameRenderer, context: DrawContext, distortionStrength: Float,
        callback: CallbackInfo
    )

    companion object {
        @JvmField
        val EVENT: Event<NauseaRenderCallback> = EventFactory.createArrayBacked<NauseaRenderCallback>(
            NauseaRenderCallback::class.java,
            NauseaRenderCallback { gameRenderer: GameRenderer, context: DrawContext, distortionStrength: Float,
                                   callback: CallbackInfo? -> }
        ) { listeners: Array<NauseaRenderCallback> ->
            NauseaRenderCallback { gameRenderer: GameRenderer, context: DrawContext, distortionStrength: Float,
                                   callback: CallbackInfo ->
                for (listener in listeners) {
                    listener.onNauseaRender(gameRenderer, context, distortionStrength, callback)

                    if (callback.isCancelled) {
                        return@NauseaRenderCallback
                    }
                }
            }
        }
    }
}
